### PR TITLE
Chore: More assertions explained

### PR DESCRIPTION
**Pull request by MikaelMayer**
After reviewing carefully #1915 , I figured out we ought to add more descriptions of common assertions that Dafny verifies, especially those that most programmers are used to (i.e. division by zero, object not null) in first position.
This will increase the understanding about what Dafny will check, and can also serve as an entry point to visit relevant sections in the manual.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: docs/DafnyRef/UserGuide.md [hunk 1]
```diff
@@ -319,17 +319,45 @@ This list is not exhaustive but can definitely be useful to provide the next ste
...
-* Every call to a function or method with a [`requires` clause](#sec-requires-clause) yields _one assertion per requires clause_[^precision-requires-clause]
-  (special cases such as sequence indexing come with a special require clause that the index is within bounds).
-* Assignments `o.f := E;` yield an _assertion_ that `o.f` is allowed by the enclosing [`modifies` clause](#sec-loop-framing).
-* [Assign-such-that operators](#sec-update-and-call-statement) `x :| P(x)` yield an _assertion_ that `exists x :: P(x)`.
-* Every [`ensures` clause](#sec-ensures-clause) yields an _assertion_ at the end of the method and on every return.
-* Every [`decreases` clause](#sec-decreases-clause) yields an _assertion_ at either a call site or at the end of a while loop.
+**Integer assertions:**
+
+* Every [division](#sec-numeric-types) yields an _assertion_ that the divisor is never zero.
+* Every [bounded number operation](#sec-numeric-types) yields an _assertion_ that the result will have the same bounds (no overflow, no underflows).
```
Comment (atomb) [inline docs/DafnyRef/UserGuide.md:327]:
```suggestion
* Every [bounded number operation](#sec-numeric-types) yields an _assertion_ that the result will be within the same bounds (no overflow, no underflows).
```

File: docs/DafnyRef/UserGuide.md [hunk 2]
```diff
@@ -319,17 +319,45 @@ This list is not exhaustive but can definitely be useful to provide the next ste
...
+* Every [conversion](#sec-as-expression) yields an _assertion_ that conversion is compatible
```
Comment (atomb) [inline docs/DafnyRef/UserGuide.md:328]:
```suggestion
* Every [conversion](#sec-as-expression) yields an _assertion_ that conversion is compatible.
```

File: docs/DafnyRef/UserGuide.md [hunk 3]
```diff
@@ -319,17 +319,45 @@ This list is not exhaustive but can definitely be useful to provide the next ste
...
+* Every [bitvector shift](#sec-bit-vector-types) yields an _assertion_ that the shift amount is never negative, and that the shift amount is within the width of the value.
+
+**Object assertions:**
+
+* Every [object property access](#sec-class-types) yields an _assertion_ that the object is not null
```
Comment (atomb) [inline docs/DafnyRef/UserGuide.md:333]:
```suggestion
* Every [object property access](#sec-class-types) yields an _assertion_ that the object is not null.
```

File: docs/DafnyRef/UserGuide.md [hunk 4]
```diff
@@ -319,17 +319,45 @@ This list is not exhaustive but can definitely be useful to provide the next ste
...
+* Every assignment `o.f := E;` yield an _assertion_ that `o` is among the set of objects of the enclosing [`modifies`](#sec-loop-framing) clause in loops, or the enclosing [`modifies`](#sec-modifies-clause) clause on methods.
```
Comment (atomb) [inline docs/DafnyRef/UserGuide.md:334]:
```suggestion
* Every assignment `o.f := E;` yield an _assertion_ that `o` is among the set of objects of the `modifies` clause of the enclosing [loop](#sec-loop-framing) or [method](#sec-modifies-clause).
```

File: docs/DafnyRef/UserGuide.md [hunk 5]
```diff
@@ -319,17 +319,45 @@ This list is not exhaustive but can definitely be useful to provide the next ste
...
+* Every read `o.f` yield an _assertion_ that `o` is among the set of objects of the enclosing [`reads`](#sec-reads-clause) clause on functions and predicates, or the enclosing [`modifies`](#sec-modifies-clause) clause on methods
```
Comment (atomb) [inline docs/DafnyRef/UserGuide.md:335]:
```suggestion
* Every read `o.f` yield an _assertion_ that `o` is among the set of objects of the [`reads`](#sec-reads-clause) clause of the function, predicate, or method; or the enclosing [`modifies`](#sec-modifies-clause) clause of the method.
```

Comment (atomb) [reply docs/DafnyRef/UserGuide.md:335]: I'm not 100% sure that my suggested change is true. We should investigate.

File: docs/DafnyRef/UserGuide.md [hunk 6]
```diff
@@ -319,17 +319,45 @@ This list is not exhaustive but can definitely be useful to provide the next ste
...
+* Every [array access](#sec-array-types) `a[x]` yield the assertion that `0 <= x < a.Length`
```
Comment (atomb) [inline docs/DafnyRef/UserGuide.md:336]:
```suggestion
* Every [array access](#sec-array-types) `a[x]` yield the assertion that `0 <= x < a.Length`.
```

File: docs/DafnyRef/UserGuide.md [hunk 7]
```diff
@@ -319,17 +319,45 @@ This list is not exhaustive but can definitely be useful to provide the next ste
...
+* Every [sequence access](#sec-sequences) `a[x]` yield an _assertion_, that `0 <= x < |a|`, because sequences are never null.
+* Every [datatype update expression](#sec-datatype-update-suffix) and [datatype destruction](#sec-algebraic-datatype) yields an _assertion_ that the object has the given property.
+* Every method overriding a [`trait`](#sec-trait-types) yield an _assertion_ that it provides a postcondition equal to or more detailed than in its parent trait, and an _assertion_ that it provides a precondition equal to or more permissive than in its parent trait.
```
Comment (atomb) [inline docs/DafnyRef/UserGuide.md:339]:
```suggestion
* Every method overriding a [`trait`](#sec-trait-types) yield an _assertion_ that any postcondition it provides is equal to or more detailed than in its parent trait, and an _assertion_ that any precondition it provides is equal to or more permissive than in its parent trait.
```

File: docs/DafnyRef/UserGuide.md [hunk 8]
```diff
@@ -319,17 +319,45 @@ This list is not exhaustive but can definitely be useful to provide the next ste
...
+
+**Implicit assertions:**
```
Comment (atomb) [inline docs/DafnyRef/UserGuide.md:341]: The integer and object assertions earlier are also implicit. We could perhaps use the headings "Implicit integer assertions", "Implicit object assertions", and "Other implicit assertions". What do you think?

**Review by atomb — CHANGES_REQUESTED** [review]
This is something very useful to have explicitly documented! I made some minor wording and consistency suggestions, but I think it's overall great.

File: docs/DafnyRef/UserGuide.md (hunk 5 above)
Comment (MikaelMayer) [reply docs/DafnyRef/UserGuide.md:335]: There is no reads clause on methods (yet)

Comment (MikaelMayer) [reply docs/DafnyRef/UserGuide.md:335]: I modified your comment slightly to commit it.

File: docs/DafnyRef/UserGuide.md (hunk 8 above)
Comment (MikaelMayer) [reply docs/DafnyRef/UserGuide.md:341]:
```suggestion
**Other implicit assertions:**
```

File: docs/DafnyRef/UserGuide.md (hunk 5 above)
Comment (atomb) [reply docs/DafnyRef/UserGuide.md]: Sounds great.

File: docs/DafnyRef/UserGuide.md [hunk 9]
```diff
@@ -319,17 +319,45 @@ This list is not exhaustive but can definitely be useful to provide the next ste
...
+* Every [division](#sec-numeric-types) yields an _assertion_ that the divisor is never zero.
+* Every [bounded number operation](#sec-numeric-types) yields an _assertion_ that the result will be within the same bounds (no overflow, no underflows).
+* Every [conversion](#sec-as-expression) yields an _assertion_ that conversion is compatible.
+* Every [bitvector shift](#sec-bit-vector-types) yields an _assertion_ that the shift amount is never negative, and that the shift amount is within the width of the value.
+
+**Object assertions:**
+
+* Every [object property access](#sec-class-types) yields an _assertion_ that the object is not null.
+* Every assignment `o.f := E;` yield an _assertion_ that `o` is among the set of objects of the `modifies` clause of the enclosing [loop](#sec-loop-framing) or [method](#sec-modifies-clause).
+* Every read `o.f` yield an _assertion_ that `o` is among the set of objects of the [`reads`](#sec-reads-clause) clause of the enclosing function or predicate; or the enclosing [`modifies`](#sec-modifies-clause) clause of the enclosing method.
```
Comment (atomb) [inline docs/DafnyRef/UserGuide.md:335]:
```suggestion
* Every read `o.f` yield an _assertion_ that `o` is among the set of objects of the [`reads`](#sec-reads-clause) clause of the enclosing function or predicate; or the [`modifies`](#sec-modifies-clause) clause of the enclosing method.
```

**Review by atomb — CHANGES_REQUESTED** [review]
One tiny additional change: an unnecessarily duplicated word.

**Review by atomb — APPROVED** [review]
🎉

**Pull request merged**